File: source/downloads/code/LintCode/Partition-List.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param x: an integer
    @return: a ListNode
    """
    def partition(self, head, x):
        # write your code here
        if head == None or head.next == None:
            return head

        dummy = ListNode(0,None)
        left_list = dummy
        dummy2 = ListNode(0,None)
        right_list = dummy2

        while head != None:
            if head.val < x:
                left_list.next = head
                left_list = left_list.next
            else:
                right_list.next = head
                right_list = right_list.next
            head = head.next
        # connect left_list and right_list
        left_list.next = dummy2.next
        # terminate the right list
        right_list.next = None
        return dummy.next

File: downloads/code/LeetCode/Java/Partition-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode less = new ListNode(Integer.MAX_VALUE);
        ListNode nLess  = new ListNode(Integer.MIN_VALUE);
        ListNode ret = less;
        ListNode nLessHead = nLess;
        
        while(head!=null){
            ListNode nextNode = head.next;
            head.next = null;
            if(head.val < x){
                less.next = head;
                less = less.next;
            }
            else{
                nLess.next = head;
                nLess = nLess.next;
            }
            head = nextNode;
        }
        less.next = nLessHead.next;
        return ret.next;
        
    }
}
